File: fixtures/scale/scala/medium/src/calc/eval/FunctionRegistry.scala
package calc.eval

import calc.errors.EvalException

final class FunctionRegistry {
  private var unary = Map.empty[String, Double => Double]
  private var aggregate = Map.empty[String, Array[Double] => Double]

  def addUnary(name: String, fn: Double => Double): Unit =
    unary = unary.updated(name, fn)

  def addAggregate(name: String, fn: Array[Double] => Double): Unit =
    aggregate = aggregate.updated(name, fn)

  def invoke(name: String, args: Array[Double]): Double = {
    if (args.length == 1 && unary.contains(name)) unary(name)(args(0))
    else aggregate.get(name) match {
      case Some(fn) => fn(args)
      case None => throw new EvalException("unknown function " + name + "/" + args.length)
    }
  }

  def knows(name: String): Boolean = unary.contains(name) || aggregate.contains(name)
}

File: fixtures/scale/kotlin/medium/src/calc/eval/FunctionRegistry.kt
package calc.eval

import calc.errors.EvalException

class FunctionRegistry {
    private val unary = mutableMapOf<String, (Double) -> Double>()
    private val aggregate = mutableMapOf<String, (DoubleArray) -> Double>()

    fun addUnary(name: String, fn: (Double) -> Double) {
        unary[name] = fn
    }

    fun addAggregate(name: String, fn: (DoubleArray) -> Double) {
        aggregate[name] = fn
    }

    fun invoke(name: String, args: DoubleArray): Double {
        if (args.size == 1) unary[name]?.let { return it(args[0]) }
        val fn = aggregate[name] ?: throw EvalException("unknown function " + name + "/" + args.size)
        return fn(args)
    }

    fun knows(name: String): Boolean = name in unary || name in aggregate
}
